src/types/bytes.rs

```diff
@@ -191,6 +214,106 @@ impl<I: SliceIndex<[u8]>> Index<I> for Bound<'_, PyBytes> {
     }
 }
 
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<[u8]> for Bound<'_, PyBytes> {
+    #[inline]
+    fn eq(&self, other: &[u8]) -> bool {
+        self.as_borrowed() == *other
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<&'_ [u8]> for Bound<'_, PyBytes> {
+    #[inline]
+    fn eq(&self, other: &&[u8]) -> bool {
+        self.as_borrowed() == **other
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<Bound<'_, PyBytes>> for [u8] {
+    #[inline]
+    fn eq(&self, other: &Bound<'_, PyBytes>) -> bool {
+        *self == other.as_borrowed()
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<&'_ Bound<'_, PyBytes>> for [u8] {
+    #[inline]
+    fn eq(&self, other: &&Bound<'_, PyBytes>) -> bool {
+        *self == other.as_borrowed()
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<Bound<'_, PyBytes>> for &'_ [u8] {
+    #[inline]
+    fn eq(&self, other: &Bound<'_, PyBytes>) -> bool {
+        **self == other.as_borrowed()
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<[u8]> for &'_ Bound<'_, PyBytes> {
+    #[inline]
+    fn eq(&self, other: &[u8]) -> bool {
+        self.as_borrowed() == other
+    }
+}
+
+/// Compares whether the Python bytes object is equal to the [u8].
+///
+/// In some cases Python equality might be more appropriate; see the note on [`PyBytes`].
+impl PartialEq<[u8]> for Borrowed<'_, '_, PyBytes> {
+    #[inline]
+    fn eq(&self, other: &[u8]) -> bool {
+        self.to_cow().map_or(false, |s| s == other)
```

Comment:
From what I can tell, all the other `impl`s are just permutations of different types that ultimate call this `impl`. I have two questions here:

1. I copied this from `types/string.py`. But I removed the special implementation for 3.10 or ABI. I assume that is specific to the string implementation. Is that correct? Or is there an analogous implementation for bytes that is special for 3.10 or ABI?
2. I'm pretty sure `to_cow()` isn't a thing for `PyBytes`. I will work on an implementation here. Do I manually map (or loop) over the two parameters and compare each element? Or is there something more idiomatic in Rust?

Reply:
I think you can use `self.as_bytes() == other` here (which should then use the `PartialEq` impl on slices on `std`)

Reply:
@Icxolu Thanks for the feedback. I tried your suggestion and tests fail with

> no implementation for `instance::Bound<'_, bytes::PyBytes> == &[u8; 12]

I guess we need the length of the array as part of the type? Or should I use `Vec<u8>` instead?

Reply:
Ah, this is because `let b = b"hello, world";` is creating an array (`&[u8, 12]`), not a slice (`&[u8]`). To test these impl you can turn that array into a slice like this: `let b = b"hello, world".as_slice();`

If we want we can also add implementations for arrays directly as well, using const generics, something like the following:
```rust
impl <const N: usize> PartialEq<[u8; N]> for Borrowed<'_, '_, PyString> { ... }
```

Reply:
...or are my tests wrong and need changed?

Reply:
They currently test comparison with arrays (which are nor implemented) instead of the comparison with slices. This can be a bit surprising, since often arrays can be automatically coerced to slices by the compiler. In this case the compiler is not smart enough to see that this is what we want.

If you change your test to use `let b = b"hello, world".as_slice();` or `let b: &[_] = b"hello, world";`, we explicitly specify that we want `b` to be a slice as opposed to an array. This way it should then also use the `PartialEq` impl that you have implemented here.

Reply:
For context, your first response wasn't showing when I typed my second comment.